### DISCUSSION TITLE
System won't solve unless I have very high nl_rel_tol

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#17809] [Q&A Modules: Solid mechanics] [answered]
Hi guys, 

My model won't solve unless I set `nl_rel_tol` to something like 0.9, I tried to increase `nl_abs_tol` and  `l_abs_tol` instead, but they don't help.

> error message:
> Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
>  Solve Did NOT Converge!

I'm running a quasi-static mining simulation with E of 20e9, rapid strain softening (DP model) with the domain size of like 3000^3m, hence the max stress is around 8e7 Pa.

is there any way to work around nl_rel_tol value? I see the coal mining example coverts everything to MPa instead of Pa, will this help? @WilkAndy 

Kind regards,
Traiwit

**Comment by WilkAndy** [comment, marked as answer]
Working in MPa won't help - it simply scales all the residuals and makes your input file (perhaps) easier to read.

Are you using Newton or PJFNK or something else?

**Reply by Traiwit** [reply]
I'm using Newton now, will give PJFNK a go though it is much slower (imo) compared to Newton

**Reply by WilkAndy** [reply]
Newton is better, i agree.  I think PJFNK will likely bomb out really early, as the model will go into some strange configuration where the plasticity goes nuts.

**Reply by Traiwit** [reply]
Yes, I tried PJFNK, didn't work so well + very slow.
all the simulation works fine if I have nl_rel_tol>0.9 with max_disp of like 2-3 cm (after stopes excavation) for a huge mine model.

I just feel like nl_rel_tol>0.9 is bad, since most of the example has this value around 1e-3, but if nothing is too wrong about the results, I might have to live with it.

Thanks @WilkAndy

**Reply by GiudGiud** [reply]
You should not accept a nl_rel_tol of 0.9. The simulation has not converged. 

Try to output the results at two different unconverged step and see if you can see any pattern.
Also use the `show_residual_norms` option to see which variable is lagging behind in convergence.

**Reply by Traiwit** [reply]
Thanks @GiudGiud, basically the simulation moves to the next step without having to solve the next non-linear iteration when I use nl_rel_tol of 0.9. I turn off plasticity and see the disp_ and stress results are pretty much 30% different.

where do I use 'show_residual_norms'? i couldn't find it in any example.

also is there anywhere else that I can relax simulation so I can reduce nl_rel_tol a little bit more?

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
With a tolerance of 0.9 your solution could be 90% different and the solver would not care. 

It's a debug output and I mispelled it
```
[Debug]
  show_var_residual_norms = true
[]
```

**Reply by GiudGiud** [reply]
Any updates on this?

With regards to relaxing the simulation, you could a fixed point iteration process, and use a relaxation factor there.

**Reply by Traiwit** [reply]
Hi @GiudGiud and @WilkAndy 

so this error 

> error message:
> Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
> Solve Did NOT Converge!

is solved by adding `use_displaced_mesh = true` into every object under the [Kernels] and [UserObject] blocks

Now simulation solves fine with rapid plastic softening (1/10 coh within 1% damage), but just take much much longer for simulation to converge for `nl_rel_tol` of less than 0.9 (0.3-0.5).

now i'm looking for the way to make the simulation convereg fasting, hence i create a new thread about adaptive convergence criteria.

Anyway, thanks a lot guys,
Traiwit

**Reply by GiudGiud** [reply]
0.3 / 0.5 still seems pretty high.
What is the absolute residual?
Do you have the output of 'show_var_residual_norms' ?

**Reply by WilkAndy** [reply]
I agree with @GiudGiud : 0.3 is still really high.

- you might be almost converged anyway, and that's why @GiudGiud suggested the absolute residual.  If it's really tiny then MOOSE won't converge any more due to precision loss.  Compared with most MOOSE users, I'm a real fan of setting `nl_abs_tol` , because many PorousFlow simulations suffer from the problem of being almost converged.  You can see a small essay i wrote at https://mooseframework.inl.gov/modules/porous_flow/convergence.html
- @GiudGiud suggested `show_var_residual_norms` in a `[Debug]` block because you can see how different `Variables` might need `scaling`.   Incorrect scaling can also lead to nonconvergence, and is also mentioned in the above link.

It is difficult to conclusively say what's wrong.  In mining scenarios with difficult plasticity, i have sometimes had to add damped dynamics to get good convergence.   Have you tried that?

**Reply by Traiwit** [reply]
Hi @WilkAndy and @GiudGiud,

it really depends on the strain-softening criteria, if I want to do 1/2 coh within 5% damage, i can set `nl_rel_tol` to as low as 1e-2

I did not set `nl_abs_tol` at all, the default value is quite low. Yes, I tried `show_var_residual_norms ` my typical residual at step 0  is in the range of 1e+8, so yeh it will never reached the `nl_abs_tol`

@WilkAndy do you mean the damping block? like this?

```
[Dampers]
  [./const_damp]
    type = ConstantDamper
    damping = 0.6
  [../]
[]
```

I have tried, but it affects my results, so I stopped using it. Also, note that I am back to the Tensor-mechanics module (with no Dynamics).

Anyways, thanks guys. I will probs check out the scaling for now.

Kind regards,
Traiwit

**Reply by WilkAndy** [reply]
Regarding the damped dynamics.... I meant, add velocities and accelerations and impose critical damping (or close to) so your system becomes a truly dynamical one (with a real notion of time) instead of a quasi-static one.  Almost definitely you'll get better but much, much slower convergence.

**Reply by Traiwit** [reply]
Hi @WilkAndy @GiudGiud ,

So I tried `scaling`, it helps on converging a lot, tho it still slow if I want to set nl_rel_tol below 0.1

so I have 3 individual variables (disp_x, disp_y, disp_z), what I did I tried to make them equal, so I multiply the first 2 terms by 1e-8 and the disp_z by 2e-9.

Is there any trick or a rule on scaling? should I push them further down? right now while running the simulation, the residuals are around 1e-1 level instead of 1e+8 (with no scaling).

@WilkAndy I am trying not to go back to dynamics at this stage, changing any of the parameters there affect my simulation results quite a lot.

Thank you.

Kind regards,
Traiwit

**Reply by WilkAndy** [reply]
The scaling you're doing is only hiding the problem.  Scaling is used when the residuals of different variables are dramatically different, which is not your case.  For instance, imagine you also have a temperature variable: you might find its residual is 1E-9 lower than the disp variables, in which case MOOSE will think you've converged even if the temperature is absolutely crazy, because it's got a tiny residual no matter what.  So, you scale temperature (`scaling = 1E9`, meaning "multiply the temperature residual by 1E9") to ensure MOOSE converges to a sensible solution.   But in your case, simply scaling the `disp` variables by 1E-8 or 2E-9 is just reducing the residual without impacting the solution.

**Comment by Traiwit** [comment]
what does 'nl_rel_tol' really mean? i see the definition says '`Description:Nonlinear Relative Tolerance`', but relative to what?

**Reply by jessecarterMOOSE** [reply]
Relative to the initial nonlinear residual (0'th nonlinear |R|) for that step.

**Comment by hugary1995** [comment]
If the convergence issue only occurs after the softening stage, you can lag the internal softening variable to get good convergence.

**Reply by Traiwit** [reply]
Hi @hugary1995,

what do you mean by 'lag the internal softening variable', do you mean increase `internal_limit`?

If so that will defeat what I want to achieve, we want to do rapid softening where cohesion goes down by 1/10 with 0.5% damage.

```
  [./mc_coh]
    type = TensorMechanicsHardeningCubic
    value_0 = 5e6
    value_residual = 5e5
    internal_limit = 0.005
    use_displaced_mesh = true
  [../]
```

Kind regards,
Traiwit

**Reply by WilkAndy** [reply]
try using `execute_on = timestep_end` or `execute_on = timestep_begin` in the above

**Reply by hugary1995** [reply]
I'm a little confused. Maybe I missed some of your earlier discussion. Are you trying to use this hardening model as a softening model? I thought you were using some sort of damage plasticity model. Are you referring to your effective plastic strain as "damage"?
